e2e-tests/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+use std::env;
+use std::fs::{create_dir_all, File};
+use std::net::TcpListener;
+use std::path::{Path, PathBuf};
+use std::process::{Child, Command, ExitStatus, Stdio};
+use std::time::Duration;
+
+use tokio::net::TcpStream;
+use url::Url;
+
+const MIN_PORT: u16 = 49_152;
+const MAX_PORT: u16 = 65_535;
+const CONNECTION_ATTEMPTS: usize = 360;
+const CONNECTION_ATTEMPT_DELAY_MS: u64 = 500;
+
+#[derive(Debug)]
+pub struct Orchestrator {
+    process: Child,
+    address: String,
+}
+
+impl Drop for Orchestrator {
+    fn drop(&mut self) {
+        let mut kill =
+            Command::new("kill").args(["-s", "TERM", &self.process.id().to_string()]).spawn().expect("Failed to kill");
+        kill.wait().expect("Failed to kill the process");
+    }
+}
+
+fn get_free_port() -> u16 {
+    for port in MIN_PORT..=MAX_PORT {
+        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
+            return listener.local_addr().expect("No local addr").port();
+        }
+        // otherwise port is occupied
+    }
+    panic!("No free ports available");
+}
+
+fn get_repository_root() -> PathBuf {
+    let manifest_path = Path::new(&env!("CARGO_MANIFEST_DIR"));
+    let repository_root = manifest_path.parent().expect("Failed to get parent directory of CARGO_MANIFEST_DIR");
+    repository_root.to_path_buf()
+}
+
+impl Orchestrator {
+    fn cargo_run(root_dir: &Path, binary: &str, args: Vec<&str>, envs: Vec<(&str, &str)>) -> Child {
+        let arguments = [vec!["run", "--bin", binary, "--release", "--"], args].concat();
+
+        let logs_dir = Path::join(root_dir, Path::new("target/logs"));
+        create_dir_all(logs_dir.clone()).expect("Failed to create logs dir");
+
+        let stdout = Stdio::from(File::create(logs_dir.join(format!("{}-stdout.txt", binary))).unwrap());
+        let stderr = Stdio::from(File::create(logs_dir.join(format!("{}-stderr.txt", binary))).unwrap());
+
+        Command::new("cargo")
+            .stdout(stdout)
+            .stderr(stderr)
+            .envs(envs)
+            .args(arguments)
+            .spawn()
+            .expect("Could not run DSN node")
+    }
+
+    pub fn run(binary: &str, args: Vec<&str>) -> Self {
+        let port = get_free_port();
+        let address = format!("127.0.0.1:{}", port);
+        let repository_root = &get_repository_root();
+
+        std::env::set_current_dir(repository_root).expect("Failed to change working directory");
+
+        let args = [args, vec!["--rpc-address", address.as_str()]].concat();
```

Comment:
do we take command line args right now?

Reply:
Nope, let me swap that with envs instead